source/endpoints_eval_util.c

```diff
@@ -0,0 +1,420 @@
+     * and % be uri encoded as %25.
+     * Implementation is fairly trivial and just iterates through the string
+     * keeping track of the spec above.
+     */
+
+    uint8_t *zone_delim = memchr(host.ptr, '%', host.len);
+
+    if (zone_delim != NULL) {
+        size_t zone_len = host.len - (size_t)(zone_delim - host.ptr);
+
+        if (zone_len < (size_t)(is_uri_encoded ? 4 : 2)) {
+            return false;
+        }
+
+        if (is_uri_encoded && (zone_delim[1] != '2' || zone_delim[2] != '5' || !aws_isalnum(zone_delim[3]))) {
+            return false;
+        }
+
+        if (!is_uri_encoded && !aws_isalnum(zone_delim[1])) {
+            return false;
+        }
+
+        host.len -= zone_len;
+    }
+
+    for (size_t i = 0; i < host.len; ++i) {
+        if (!aws_isxdigit(host.ptr[i]) && host.ptr[i] != ':') {
+            return false;
+        }
+    }
+
+    bool has_double_colon = false;
+    uint8_t colon_count = 0;
+
+    uint8_t *colon_delim = memchr(host.ptr, ':', host.len);
+    while (colon_delim != NULL) {
+        size_t group_len = colon_delim - host.ptr;
+        if (group_len == 0) {
+            if (has_double_colon) {
+                return false;
+            }
+            has_double_colon = true;
+        } else {
+            ++colon_count;
+        }
+
+        if (group_len > 4) {
+            return false;
+        }
+
+        aws_byte_cursor_advance(&host, group_len + 1);
+        colon_delim = memchr(host.ptr, ':', host.len);
+    }
+
+    if (host.len == 0 || host.len > 4) {
+        return false;
+    }
+
+    return has_double_colon ? colon_count < 6 : colon_count == 7;
+}
+
+static char s_known_countries[][3] = {{"us"}, {"eu"}, {"ap"}, {"sa"}, {"ca"}, {"me"}, {"af"}};
+
+struct aws_byte_cursor aws_map_region_to_partition(struct aws_byte_cursor region) {
+    if (region.len > AWS_REGION_LEN - 1) {
+        AWS_LOGF_ERROR(
+            AWS_LS_SDKUTILS_ENDPOINTS_GENERAL,
+            "Unexpected length of region string: " PRInSTR,
+            AWS_BYTE_CURSOR_PRI(region));
+        return aws_byte_cursor_from_c_str("");
+    }
+
+    char copy[AWS_REGION_LEN] = {0};
+    memcpy(copy, region.ptr, region.len);
+
+    char country[3] = {0};
+    char location[31] = {0};
+    uint8_t num = 0;
+
+    if (3 == sscanf(copy, "%2[^-]-%30[^-]-%03" SCNu8, country, location, &num)) {
+        for (size_t i = 0; i < sizeof(s_known_countries) / sizeof(s_known_countries[0]); ++i) {
+            if (0 == strncmp(s_known_countries[i], country, 3)) {
+                if (location[0] != 0 && num > 0) {
+                    return aws_byte_cursor_from_c_str("aws");
+                }
+            }
+        }
+    }
+
+    if (2 == sscanf(copy, "us-gov-%30[^-]-%03" SCNu8, location, &num)) {
+        if (location[0] != 0 && num > 0) {
+            return aws_byte_cursor_from_c_str("aws-us-gov");
+        }
+    }
+
+    if (2 == sscanf(copy, "cn-%30[^-]-%03" SCNu8, location, &num)) {
```

Comment:
there is an additional check for country codes in the first matcher, so it would match on scanf level, but not on country level and fallthrough to cn match. i kept matching here close to how its defined in partitions regexes, but unifying cn and general checks are probably a cleaner approach